tests/test_end_to_end.py

```diff
@@ -0,0 +1,44 @@
+import apache_beam as beam
+import pytest
+import xarray as xr
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.testing.test_pipeline import TestPipeline
+
+from pangeo_forge_recipes.patterns import CombineOp, DimKey
+from pangeo_forge_recipes.transforms import (
+    DatasetToSchema,
+    DetermineSchema,
+    IndexItems,
+    OpenWithXarray,
+    PrepareZarrTarget,
+    StoreToZarr,
+)
+
+# from apache_beam.testing.util import assert_that, equal_to
+# from apache_beam.testing.util import BeamAssertException, assert_that, is_not_empty
+
+
+@pytest.fixture
+def pipeline():
+    options = PipelineOptions(runtime_type_check=False)
+    with TestPipeline(options=options) as p:
+        yield p
+
+
+def test_xarray_zarr(
+    daily_xarray_dataset, netcdf_local_file_pattern_sequential, pipeline, tmp_target_url
+):
+    pattern = netcdf_local_file_pattern_sequential
+    with pipeline as p:
+        inputs = p | beam.Create(pattern.items())
+        datasets = inputs | OpenWithXarray(file_type=pattern.file_type)
+        # TODO determine this dynamically
+        combine_dims = [DimKey("time", operation=CombineOp.CONCAT)]
+        schemas = datasets | DatasetToSchema()
+        schema = schemas | DetermineSchema(combine_dims=combine_dims)
+        indexed_datasets = datasets | IndexItems(schema=schema)
+        target = schema | PrepareZarrTarget(target_url=tmp_target_url)
+        _ = indexed_datasets | StoreToZarr(target_store=target)
```

Comment:
Here's a translation (really, for myself) to more idiomatic Beam:

```suggestion
       combine_dims = [DimKey("time", operation=CombineOp.CONCAT)]
       datasets = (
           p
           | beam.Create(pattern.items())
           | OpenWithXarray(file_type=pattern.file_type)
        )
        schema = (
            datasets 
            | DatasetToSchema() 
            | DetermineSchema(combine_dims=combine_dims)
        )
        target = schema | PrepareZarrTarget(target_url=tmp_target_url)
        _ = (
            datasets 
            | IndexItems(schema=schema)
            | StoreToZarr(target_store=target)
         )
```

My high level feedback is that this relies seems a bit too imperative. This is my bias, but I could see the pipeline fitting closer to the [XArray Beam abstractions](https://github.com/google/xarray-beam/blob/4dced4ead5a9511ab18027780f3a98f964584067/examples/era5_climatology.py#L70), something like:

```
(
    p
    | beam.PatternToChunks(pattern)
    # other processing...
    | beam.ChunksToZarr()
```

Though, I could see how each of these steps would be composed out of the components you have here. Exposing the right level of granularity is hard.

Reply:
From your comment below: 
> Perhaps some of the intermediate steps can be merged into a single PTransform? I'm not sure the user needs to be passing schemas around.

I totally agree with this sentiment!